### PR TITLE
bpo-40747: Make py_version_nodot 3_10 not 310 (PEP 641)

### LINKED ISSUE #308: Python 3.10: _version_nodot, interpreter_version disagree on how to format it
There is a problem building and using wheels for python 3.10 xref pypa/wheel#354 and pypa/pip#8312. I think the source of the problem is that wheel's `bdist_wheel` calls `tags.interpreter_version`. But `tags.interpreter_version` seems to be producing `310`, which it will take from `sysconfig.get_config_var("py_version_nodot")` (if it exists), and there is no check that is conformant with the `3_10` format.

Internally, packaging uses `tags._version_nodot` which will always produce `3_10`, this is the "correct" value used by pip and in `tags.sys_tags`. 

[This is the code](https://github.com/python/cpython/blob/master/Lib/sysconfig.py#L89) that generates [that value](https://github.com/python/cpython/blob/master/Lib/sysconfig.py#L534) for CPython master. Note there is no '_'

### PR DESCRIPTION
**Pull request by zooba**
Implement PEP 641, and fixes pypa/packaging#308, pypa/wheel#354, and maybe pypa/pip#8312. The current value of `310` is ambiguous so use `3_10` or `3.10` as appropriate.

Also cleans up some woefully out-of-date documentation.

Co-authored-by: mattip <matti.picus@gmail.com>

https://bugs.python.org/issue40747

### CODE REVIEW

File: PC/pyconfig.h
Comment (mattip) [inline PC/pyconfig.h:272]: Wouldn’t it be better to consistently use `_` everywhere? Then no one has to think which to use.

File: PCbuild/pcbuild.proj
Comment (brettcannon) [inline PCbuild/pcbuild.proj]: I thought this was going to be `python3.X.dll` according to the doc changes in the rest of this PR?

File: PCbuild/python.props
Comment (brettcannon) [inline PCbuild/python.props]: Same as above; the rest of this PR suggests this would be `python3.X.dll`.

File: PC/pyconfig.h
Comment (zooba) [reply PC/pyconfig.h:272]: Other platforms already use dots in the equivalent filename. We decided on consistency there (though if I can't fix the crashes, we may roll back this part entirely and just change packaging stuff).

**Review by ericsnowcurrently — APPROVED** [review]
LGTM (I guess 😄)

Nothing jumped out at me after a quick read through the changes.  However, I am not very familiar with this stuff.  (Presumably you've exercised all this by hand and it works.)  Also, I haven't checked for anything in the code base you might have missed (but am not sure what I'd look for).

If you can find someone more familiar with this stuff you might want another opinion before merging. 😄

**Comment by zooba** [comment]
Thanks Eric! There's still a PEP to approve, which will attract more eyes I'm sure (plus a discussion at the sprints about whether it's worth changing everything away from "310" or not).

File: Doc/extending/windows.rst
Comment (nulano) [inline Doc/extending/windows.rst]: Are these supposed to be `pythonX.Y.lib` and `pythonX.Y.dll`?

**Comment by encukou** [comment]
It seems not all instances of `310` are actually replaced -- when I built and installed Python with the PR, I still got filenames like `__pycache__/os.cpython-310.pyc`.

**Comment by brettcannon** [comment]
@encukou I honestly don't think the `.pyc` file names need updating (although they can be if people want to put in the effort). That is entirely an implementation detail that no normal person should ever look at or care about.

**Comment by zooba** [comment]
>  I honestly don't think the .pyc file names need updating (although they can be if people want to put in the effort). That is entirely an implementation detail that no normal person should ever look at or care about.

What is the proposed tag for v31.0 then? There are a number of other names that qualify as an implementation detail, so we should be consistent about updating all or none.

Another alternative to this whole patch is to just override the wheel tag generation code (wherever that is...), rather than changing any of the internal variables at all.

**Comment by brettcannon** [comment]
@zooba generic interpreter tag version calculation:

https://github.com/pypa/packaging/blob/c4e10743c37e33cfe7b0eafc5138b5cff578979e/packaging/tags.py#L782-L793

**Comment by hroncok** [comment]
> Another alternative to this whole patch is to just override the wheel tag generation code (wherever that is...)

https://github.com/pypa/packaging/pull/354

**Comment by zooba** [comment]
The PEP has been rejected.

**Pull request closed without merging**
